Descent3/mission_download.cpp

```diff
@@ -338,181 +334,161 @@ int msn_DownloadWithStatus(char *url, char *filename) {
 
-                  time_elapsed = timer_GetTime()-starttime;
+  std::string download_uri = "/" + StringJoin(std::vector<std::string>(url_parts.begin() + 3, url_parts.end()), "/");
 
-                  if(total_bytes)
-                  {
-                          time_remain = ((float)(total_bytes-received_bytes))/((float)(received_bytes/time_elapsed));
-                  }
-                  if(time_elapsed&&received_bytes)
-                  {
-                          xfer_rate = ((float)(received_bytes/time_elapsed));
-                  }
-                  texts[1].Destroy();
-                  texts[2].Destroy();
-                  texts[3].Destroy();
-                  texts[4].Destroy();
-                  texts[5].Destroy();
-                  texts[6].Destroy();
+  while (!exit_menu) {
+    int res;
 
-                  sprintf(fmturl,DOWNLOAD_STATUS_URL_TEXT,url);
-                  msn_ClipURLToWidth(MSN_DWNLD_STATUS_W-(MSN_COL_1+MSN_BORDER_W),fmturl);
-                  sprintf(fmtrcvd,DOWNLOAD_STATUS_RCVD_TEXT,received_bytes);
-                  sprintf(fmttotal,DOWNLOAD_STATUS_TOTAL_TEXT,total_bytes);
-                  sprintf(fmtelaps,DOWNLOAD_STATUS_ELAPS_TEXT,msn_SecondsToString(time_elapsed));
-                  sprintf(fmttimer,DOWNLOAD_STATUS_TIME_R_TEXT,msn_SecondsToString(time_remain));
-                  msn_ClipURLToWidth(MSN_DWNLD_STATUS_W-(MSN_COL_2+MSN_BORDER_W),fmttimer);
-                  sprintf(fmtrate,DOWNLOAD_STATUS_XFERRATE_TEXT,xfer_rate);
+    std::fstream in;
+    in.open(qualfile, std::ios::binary | std::ios::trunc | std::ios::out);
+    int status = http_client->Get(download_uri, in, [&received_bytes, &total_bytes](uint64_t len, uint64_t total) {
+      received_bytes = len;
+      total_bytes = total;
+      return true;
+    });
+    in.close();
+
+    if ((timer_GetTime() - last_refresh) > MSN_REFRESH_INTERVAL) {
+
+      if (status == 200) {
+        // File transfer successful!
+        LOG_INFO << "Successfully received the file!";
+        exit_menu = 1;
+
+        if (file_is_zip) {
+          // now we gotta handle the zip file
+          ret = msn_ExtractZipFile(qualfile.u8string().c_str(), filename);
+        } else {
+          ret = 1;
+        }
+      }
 
-                  download_text = UITextItem (fmturl,UICOL_TEXT_NORMAL);;
-                  rcvd_text = UITextItem (fmtrcvd,UICOL_TEXT_NORMAL);
-                  total_text = UITextItem (fmttotal,UICOL_TEXT_NORMAL);
-                  elaps_text = UITextItem (fmtelaps,UICOL_TEXT_NORMAL);
-                  time_r_text = UITextItem (fmttimer,UICOL_TEXT_NORMAL);
-                  rate_text = UITextItem (fmtrate,UICOL_TEXT_NORMAL);
+      if (status != 200) {
+        // File transfer Error!
+        DoMessageBox(TXT_ERROR, TXT_FMTCANTDNLD, MSGBOX_OK);
+        // Delete the file that didn't finish!
+        std::filesystem::remove(qualfile);
+        LOG_WARNING.printf("Couldn't download the file! Error: %d", status);
```

Comment:
The server likely also sent a message in the response payload that specifies the error. We may also log it and put it in the modal dialog window

Reply:
Showing error in windows requires to mess with localized messages, I'd like not touching them yet. I left TODOs for later.